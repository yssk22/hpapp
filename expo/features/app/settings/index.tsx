import { logLogin, logLogout } from '@hpapp/system/firebase';
import * as logging from '@hpapp/system/logging';
import { useCallback, useEffect, useMemo, useState } from 'react';
import { Platform } from 'react-native';

import SettingsAppConfig, { AppConfig, SettingsAppConfigDefault } from './internals/SettingsAppConfig';
import SettingsCurrentUser, { CurrentUser } from './internals/SettingsCurrentUser';
import SettingsList from './internals/SettingsList';
import Provider, { useSettings } from './internals/SettingsProvider';
import SettingsUPFCConfig, { UPFCConfig } from './internals/SettingsUPFCConfig';
import SettingsUserConfig, { UserConfig, SettingsUserConfigDefault } from './internals/SettingsUserConfig';

export type SettingsProviderProps = {
  appConfig?: AppConfig;
  userConfig?: UserConfig;
  currentUser?: CurrentUser;
  upfcConfig?: UPFCConfig;
};

/**
 * A provider to setup the settings context
 */
export function SettingsProvider({
  appConfig,
  userConfig,
  currentUser,
  upfcConfig,
  children
}: SettingsProviderProps & {
  children: React.ReactNode;
}) {
  const [ready, setReady] = useState(false);
  useEffect(() => {
    (async () => {
      if (Platform.OS === 'web') {
        // we expect web is used only for storybook so we set the user token here from the environment variable and enforce the logged in user
        const accessToken = process.env['HPAPP_USER_TOKEN_FOR_JEST'];
        if (accessToken === undefined) {
          throw new Error('HPAPP_USER_TOKEN_FOR_JEST is not set for web');
        }
        await SettingsCurrentUser.save({
          id: '1234',
          username: 'yssk22',
          accessToken
        });
      } else {
        currentUser && (await SettingsCurrentUser.save(currentUser));
        userConfig && (await SettingsUserConfig.save(userConfig));
        appConfig && (await SettingsAppConfig.save(appConfig));
        upfcConfig && (await SettingsUPFCConfig.save(upfcConfig));
      }
      setReady(true);
    })();
  }, []);
  if (!ready) {
    return null;
  }
  return <Provider settings={SettingsList}>{children}</Provider>;
}

/**
 * Get the application configuration
 */
export function useAppConfig() {
  const [appConfig] = useSettings(SettingsAppConfig);
  return useMemo(() => {
    return {
      ...SettingsAppConfigDefault,
      ...appConfig
    };
  }, [appConfig]);
}

export { SettingsAppConfigDefault };

/**
 * Get the updator for AppConfig
 */
export function useAppConfigUpdator() {
  const [, updator] = useSettings(SettingsAppConfig);
  return updator;
}

/**
 * Get the user configuration
 */
export function useUserConfig() {
  const [userConfig] = useSettings(SettingsUserConfig);
  return userConfig;
}

/**
 * Get the updator for user configuration
 */
export function useUserConfigUpdator() {
  const [, updator] = useSettings(SettingsUserConfig);
  return updator;
}

/**
 * Get the UPFC configuration
 */
export function useUPFCConfig() {
  const [upfcConfig] = useSettings(SettingsUPFCConfig);
  // migrate from legacy helloproject credentials to new one to support mline
  return useMemo(() => {
    if (upfcConfig) {
      return {
        ...upfcConfig,
        username: undefined,
        password: undefined,
        // eslint-disable-next-line deprecation/deprecation
        hpUsername: upfcConfig?.username ?? upfcConfig?.hpUsername,
        // eslint-disable-next-line deprecation/deprecation
        hpPassword: upfcConfig?.password ?? upfcConfig?.hpPassword
      };
    }
    return undefined;
  }, [upfcConfig]);
}

/**
 * Get the updator for UPFC configuration
 */
export function useUPFCConfigUpdator() {
  const [, updator] = useSettings(SettingsUPFCConfig);
  return updator;
}

/**
 * Get the current user
 */
export function useCurrentUser() {
  const [currentUser] = useSettings(SettingsCurrentUser);
  return currentUser;
}

export type User = NonNullable<ReturnType<typeof useCurrentUser>>;

/**
 * Get the updator for current user
 */
export function useCurrentUserUpdator() {
  const [currentUser, updator] = useSettings(SettingsCurrentUser);
  return useCallback(
    (update: CurrentUser | null) => {
      if (currentUser) {
        if (update == null) {
          logging.Info('features.auth.logout', 'logout', {
            userid: currentUser.id
          });
          logLogout();
        }
      } else {
        if (update) {
          logging.Info('features.auth.login', 'login', {
            userid: update.id
          });
          // TODO: we should have a way to know the login method
          logLogin('default', update.id);
        }
      }
      return updator(update);
    },
    [updator]
  );
}

export { SettingsUserConfigDefault };
