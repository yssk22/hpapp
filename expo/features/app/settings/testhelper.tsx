import { act, render } from '@testing-library/react-native';

import { SettingsProvider, SettingsProviderProps } from './';
import SettingsList from './internals/SettingsList';

/**
 * Clear all settings
 */
export async function clearSettings() {
  await Promise.all(
    SettingsList.map((settings) => {
      settings.clear();
    })
  );
}

export async function renderSettingsProvider(children: React.ReactNode, props: SettingsProviderProps = {}) {
  await clearSettings();
  const content = render(<SettingsProvider {...props}>{children}</SettingsProvider>);
  await act(async () => {});
  return content;
}
