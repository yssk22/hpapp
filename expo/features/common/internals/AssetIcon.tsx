import { Icon } from '@rneui/themed';

import AmebloIcon from './AmebloIcon';

export type AssetIconType = 'ameblo' | 'instagram' | 'twitter';

const AssetIcon: React.FC<{
  type: AssetIconType | undefined;
  size: number;
}> = ({ type, size }) => {
  switch (type) {
    case 'ameblo':
      return <AmebloIcon size={size} />;
    case 'instagram':
      return <Icon type="ionicon" name="logo-instagram" size={size} />;
    case 'twitter':
      return <Icon type="ionicon" name="logo-twitter" size={size} color="#1da1f2" />;
    default:
      return null;
  }
};

export default AssetIcon;
