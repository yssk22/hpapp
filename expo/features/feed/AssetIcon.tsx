import { HPFeedItemHPAssetType } from '@hpapp/features/feed/__generated__/FeedListItemFragment.graphql';
import AmebloIcon from '@hpapp/features/feed/ameblo/AmebloIcon';
import { Icon } from '@rneui/themed';

const AssetIcon: React.FC<{
  type: HPFeedItemHPAssetType | undefined;
  size: number;
}> = ({ type, size }) => {
  switch (type) {
    case 'ameblo':
      return <AmebloIcon size={size} />;
    case 'instagram':
      return <Icon type="ionicon" name="logo-instagram" size={size} />;
    case 'twitter':
      return <Icon type="ionicon" name="logo-twitter" size={size} color="#1da1f2" />;
    default:
      return null;
  }
};

export default AssetIcon;
