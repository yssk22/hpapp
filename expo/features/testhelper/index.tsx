import {
  mockUseNavigation,
  renderUserScreen,
  renderUserComponent,
  renderGuestComponent,
  waitForUserRootRerendered
} from './internals/jest';
import { ScreenDecorator } from './internals/storybook';

export {
  mockUseNavigation,
  renderUserScreen,
  renderUserComponent,
  renderGuestComponent,
  waitForUserRootRerendered,
  ScreenDecorator
};
