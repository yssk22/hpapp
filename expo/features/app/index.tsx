import { maybeDev } from '@hpapp/foundation/environment';
import { LocalMediaManagerProvider } from '@hpapp/system/media';

import AppRelayProvider from './internals/AppRelayProvider';
import { default as Root, AppRootProps as RootProps } from './internals/AppRoot';
import AppUpdateBanner from './internals/AppUpdateBanner';
import { ThemeProvider, ThemeProviderProps } from './theme';

export type AppRootProps = RootProps & ThemeProviderProps;

export function AppRoot({ screens, ...props }: AppRootProps) {
  return (
    <ThemeProvider {...props}>
      <AppRelayProvider>
        <LocalMediaManagerProvider name={maybeDev() ? 'hellofanapp-dev' : 'hellofanapp'}>
          <AppUpdateBanner />
          <Root screens={screens} />
        </LocalMediaManagerProvider>
      </AppRelayProvider>
    </ThemeProvider>
  );
}
