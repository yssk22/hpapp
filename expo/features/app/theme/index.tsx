import { SettingsProvider, SettingsProviderProps } from '@hpapp/features/app/settings';
import { createTheme, calcYIQ, rgbToColor, colorToRGB } from '@hpapp/system/theme/helper';
import { ThemeColorScheme, ThemeColorKey } from '@hpapp/system/theme/types';
import { useTheme } from '@rneui/themed';
import React, { useCallback } from 'react';

import Provider from './internals/ThemeProvider';

export type ThemeProviderProps = SettingsProviderProps;

/**
 * provides the theme context. It has to be under SettingsProvider.
 */
export function ThemeProvider({ children, ...props }: { children: React.ReactNode } & SettingsProviderProps) {
  return (
    <SettingsProvider {...props}>
      <Provider>{children}</Provider>
    </SettingsProvider>
  );
}

const yiqThreshod = 64;

/**
 * get the color and contrast color from the color scheme
 * @param scheme
 * @returns
 */
export function useThemeColor(scheme: ThemeColorScheme) {
  const { theme } = useTheme();
  const color = theme.colors[scheme];
  const def = rgbToColor(color);
  if (def) {
    return [color, def.contrastColor];
  }
  const yiq = calcYIQ(color);
  return [color, yiq > yiqThreshod ? theme.colors.white : theme.colors.black];
}

/**
 * get the color for the skelton
 * @param scheme
 * @param percent
 * @returns
 */
export function useThemeSkeltonColor(scheme: ThemeColorScheme, percent = 0.8) {
  const { theme } = useTheme();
  const color = theme.colors[scheme];
  const [r, g, b] = colorToRGB(color);
  const r1 = Math.min(255, Math.floor(r + (255 - r) * percent));
  const g1 = Math.min(255, Math.floor(g + (255 - g) * percent));
  const b1 = Math.min(255, Math.floor(b + (255 - b) * percent));
  // convert r1, g1, b1 to hex
  return `#${r1.toString(16).padStart(2, '0')}${g1.toString(16).padStart(2, '0')}${b1.toString(16).padStart(2, '0')}`;
}

/**
 * Get the updator function for the theme
 * @returns a function to update the theme
 */
export function useThemeUpdator() {
  const theme = useTheme();
  return useCallback(
    () =>
      function (primaryColor: ThemeColorKey, secondaryColor: ThemeColorKey, bgColor: ThemeColorKey) {
        theme.updateTheme(createTheme(primaryColor, secondaryColor, bgColor));
      },
    [theme.updateTheme]
  );
}
