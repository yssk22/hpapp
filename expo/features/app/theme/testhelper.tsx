import { SettingsProviderProps } from '@hpapp/features/app/settings';
import { renderSettingsProvider } from '@hpapp/features/app/settings/testhelper';

import { ThemeProvider } from './';

export async function renderThemeProvider(children: React.ReactNode, props: SettingsProviderProps = {}) {
  return renderSettingsProvider(<ThemeProvider>{children}</ThemeProvider>, props);
}
