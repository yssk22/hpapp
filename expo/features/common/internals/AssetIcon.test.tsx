import { renderUserComponent } from '@hpapp/features/app/testhelper';
import { composeStories } from '@storybook/react';

import * as stories from './AssetIcon.stories';

describe('AssetIcon', () => {
  const composed = composeStories(stories);
  Object.keys(composed).forEach((key) => {
    const C = (composed as any)[key];
    test(key, async () => {
      const content = await renderUserComponent(<C />);
      expect(content).toMatchSnapshot();
    });
  });
});
