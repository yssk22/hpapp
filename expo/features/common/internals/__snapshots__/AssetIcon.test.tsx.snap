// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetIcon Basic 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNSScreenStack
    style={
      {
        "flex": 1,
      }
    }
  >
    <RNSScreen
      collapsable={false}
      gestureResponseDistance={
        {
          "bottom": -1,
          "end": -1,
          "start": -1,
          "top": -1,
        }
      }
      nativeBackButtonDismissalEnabled={false}
      onAppear={[Function]}
      onDisappear={[Function]}
      onDismissed={[Function]}
      onGestureCancel={[Function]}
      onHeaderBackButtonClicked={[Function]}
      onNativeDismissCancelled={[Function]}
      onTransitionProgress={[Function]}
      onWillDisappear={[Function]}
      replaceAnimation="push"
      sheetAllowedDetents="large"
      sheetCornerRadius={-1}
      sheetExpandsWhenScrolledToEdge={true}
      sheetGrabberVisible={false}
      sheetLargestUndimmedDetent="all"
      stackPresentation="push"
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      swipeDirection="horizontal"
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        style={
          {
            "flex": 1,
            "flexDirection": "column-reverse",
          }
        }
      >
        <View
          collapsable={false}
          style={
            [
              {
                "flex": 1,
              },
              {
                "backgroundColor": "white",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              {
                "backgroundColor": "#ffffff",
                "flex": 1,
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "flex": 1,
                  "justifyContent": "center",
                }
              }
            >
              <RNSVGSvgView
                align="xMidYMid"
                bbHeight={16}
                bbWidth={16}
                focusable={false}
                meetOrSlice={0}
                minX={0}
                minY={0}
                style={
                  [
                    {
                      "backgroundColor": "transparent",
                      "borderWidth": 0,
                    },
                    {
                      "height": 16,
                      "width": 16,
                    },
                    {
                      "flex": 0,
                      "height": 16,
                      "width": 16,
                    },
                  ]
                }
                vbHeight={34.7}
                vbWidth={34.7}
                x="0px"
                y="0px"
              >
                <RNSVGGroup
                  fill={
                    {
                      "payload": 4278190080,
                      "type": 0,
                    }
                  }
                >
                  <RNSVGGroup
                    fill={
                      {
                        "payload": 4278190080,
                        "type": 0,
                      }
                    }
                  >
                    <RNSVGPath
                      d="M28.7,24.5c-2-1-1.9-3.5-1.7-5.9c0.3-2.4,0.3-3.7-0.3-5.8C26.1,11,25.1,10,25.1,8c0-1.5,0-3.5,0-4.7c0-1.8-0.8-2.4-1.6-2.4 c-2.3,0-3.1,4-6.2,4h0c-3.1,0-3.9-4-6.2-4c-0.7,0-1.6,0.6-1.6,2.4c0,1.3,0,3.3,0,4.7c0,2-1,3-1.6,4.8c-0.6,2.1-0.6,3.4-0.4,5.8 C7.9,21.1,8,23.5,6,24.5c-1.6,0.8-2.6,2.1-1.8,3.2c0.9,1.4,2.6-0.2,3.4,0c0.9,0.2-0.1,1.5,1.3,1.5c0.7,0,0.9-2.2,1.6-2 c0.5,0.1,0.6,1.5,0.1,2.5c-0.2,0.5-2,1.6-1.8,2.8c0.2,1.2,1.2,1.4,2.6,1.4c1,0,2.7,0,3.7,0c0.9,0,1.8-0.9,1.8-1.5 c0-0.6-0.8-1.9,0.5-1.9h0c1.3,0,0.5,1.3,0.5,1.9c0,0.6,0.9,1.5,1.8,1.5c0.9,0,2.6,0,3.7,0c1.4,0,2.4-0.2,2.6-1.4 c0.2-1.2-1.6-2.3-1.8-2.8c-0.5-1.1-0.3-2.4,0.1-2.5c0.7-0.2,0.9,2,1.6,2c1.4,0,0.4-1.3,1.3-1.5c0.8-0.2,2.5,1.4,3.4,0 C31.3,26.6,30.3,25.3,28.7,24.5z"
                      fill={
                        {
                          "payload": 4278222848,
                          "type": 0,
                        }
                      }
                      propList={
                        [
                          "fill",
                        ]
                      }
                    />
                  </RNSVGGroup>
                  <RNSVGCircle
                    cx="22.4"
                    cy="11.5"
                    fill={
                      {
                        "payload": 4294967295,
                        "type": 0,
                      }
                    }
                    propList={
                      [
                        "fill",
                      ]
                    }
                    r="0.9"
                  />
                  <RNSVGCircle
                    cx="12.3"
                    cy="11.5"
                    fill={
                      {
                        "payload": 4294967295,
                        "type": 0,
                      }
                    }
                    propList={
                      [
                        "fill",
                      ]
                    }
                    r="0.9"
                  />
                  <RNSVGCircle
                    cx="17.4"
                    cy="25.1"
                    fill={
                      {
                        "payload": 4294967295,
                        "type": 0,
                      }
                    }
                    propList={
                      [
                        "fill",
                      ]
                    }
                    r="4.2"
                  />
                  <RNSVGPath
                    d="M17.4,12.2c-2.6,0-6.2,1.6-6.2,4.1c0,2,1.3,2.7,2.9,2.7s2.1-1.8,3.4-1.8c1.3,0,1.8,1.8,3.4,1.8 c1.6,0,2.9-0.7,2.9-2.7C23.6,13.8,20,12.2,17.4,12.2z M17.4,15.9c-1.1,0-2.1-0.7-2.1-1.5c0-0.9,0.9-1.5,2.1-1.5 c1.1,0,2.1,0.7,2.1,1.5C19.4,15.2,18.5,15.9,17.4,15.9z"
                    fill={
                      {
                        "payload": 4294967295,
                        "type": 0,
                      }
                    }
                    propList={
                      [
                        "fill",
                      ]
                    }
                  />
                </RNSVGGroup>
              </RNSVGSvgView>
            </View>
          </View>
        </View>
      </View>
      <RNSScreenStackHeaderConfig
        backButtonInCustomView={false}
        backTitle=""
        backTitleVisible={false}
        backgroundColor="#0075c2"
        color="white"
        direction="ltr"
        disableBackButtonMenu={false}
        hidden={false}
        hideBackButton={false}
        largeTitleHideShadow={false}
        title="/"
        titleColor="white"
        topInsetEnabled={false}
        translucent={false}
      />
    </RNSScreen>
  </RNSScreenStack>
</RNCSafeAreaProvider>
`;
